### PR TITLE
Add local echo for notifications in the new room list

### PR DESCRIPTION
**Pull request by turt2live**
# Design review portion

Is this copy good enough?

image

... then when the user manages to `resend` everything ...

image

-----

# Code review portion

For https://github.com/vector-im/riot-web/issues/14302
Fixes https://github.com/vector-im/riot-web/issues/14280

This is somewhat reviewable commit-by-commit, though I'd instead recommend reading the included `local-echo-dev.md` docs and then reviewing the code.

Live demo:
notification-echo mp4

TODO:
* [x] Wire up a UI for failures
* [x] Validate/document the class structure

### CODE REVIEW

**Comment by turt2live** [comment]
Copy reviewed oob

File: src/stores/ToastStore.ts [hunk 1]
```diff
@@ -55,7 +56,7 @@ export default class ToastStore extends EventEmitter {
      *
      * @param {object} newToast The new toast
      */
-    addOrReplaceToast<C extends keyof JSX.IntrinsicElements | JSXElementConstructor<any>>(newToast: IToast<C>) {
+    addOrReplaceToast<C extends ComponentClass>(newToast: IToast<C>) {
```
Comment (jryans) [inline src/stores/ToastStore.ts:59]: I don't think this type was correct before, and it also does not seem the same as your new store (which is sharing the type at the moment) either...?

This older store seems to take an object of several things, including component class, props, etc. It manually peels off the `component` key, and hands that `createElement`. Your new store expects a single component class directly, which seems quite different, but maybe I am missing something...?

File: src/stores/NonUrgentToastStore.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            NonUrgentToastStore._instance = new NonUrgentToastStore();
+        }
+        return NonUrgentToastStore._instance;
+    }
+
+    public get components(): ComponentClass[] {
+        return Array.from(this.toasts.values());
+    }
+
+    public addToast(c: ComponentClass): ToastReference {
```
[the remark this answers is not in the document]
Comment (jryans) [reply src/stores/NonUrgentToastStore.ts:39]: (I'm not trying to say anything needs to be changed here, just reflecting on what might be more natural to use.)

**Comment by jryans** [comment]
Since this is relatively isolated to notification state for the moment, it seems safe enough to land and try for the upcoming RC, so I'll merge this now. Please do look over the comments above and consider fixing after the fact where needed.

**Comment by turt2live** [comment]
> I'm not quite sure why you've marked this as needing to be in the RC though...? It seems like a new feature that can merge at any time.

It fixes a regression where the notifications menu appears to do nothing. We could in theory try and fix it, but notifications are so badly scoped that either the work would be undone by ftue notifications or we'd regress performance. Local echo is just one of the ways out of both problems.

**Comment by niquewoodhouse** [comment]
lgtm for first iteration

File: src/stores/NonUrgentToastStore.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/NonUrgentToastStore.ts:39]: yea, it's a bit awkward how we create all this stuff. Something to review for toasts in general.

File: src/stores/local-echo/EchoTransaction.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+*/
+
+import { Whenable } from "../../utils/Whenable";
+
+export type RunFn = () => Promise<void>;
+
+export enum TransactionStatus {
+    Pending,
+    DoneSuccess,
+    DoneError,
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/local-echo/EchoTransaction.ts:24]: This is somewhat intentional to indicate that the context has settled

File: src/stores/local-echo/GenericEchoChamber.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        // Cancel any pending transactions for the same key
+        if (this.cache.has(key)) {
+            this.cache.get(key).txn.cancel();
+        }
+
+        const txn = this.context.beginTransaction(auditName, runFn);
+        this.cacheVal(key, targetVal, txn); // set the cache now as it won't be updated by the .when() ladder below.
+
+        txn.when(TransactionStatus.Pending, () => this.cacheVal(key, targetVal, txn))
+            .when(TransactionStatus.DoneError, () => revertFn());
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/local-echo/GenericEchoChamber.ts:87]: our style rules say 4 space indent, as much as I hate it for cases like this

File: src/stores/ToastStore.ts (hunk 1 above)
Comment (turt2live) [reply src/stores/ToastStore.ts:59]: it's the right type, just looks really strange. The existing toasts do still take a bunch of properties via the `IToast` class, which is generified to take a `ComponentClass`. The `<C extends ComponentClass>` is more or less a type definition so we can use `C` later in the function's definition.

File: src/stores/local-echo/EchoTransaction.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jryans) [reply src/stores/local-echo/EchoTransaction.ts:24]: I guess it feels like new terms on top of promises for the same concepts... Maybe "fulfilled" and "rejected" from promises could be used?

File: src/stores/local-echo/GenericEchoChamber.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jryans) [reply src/stores/local-echo/GenericEchoChamber.ts:87]: Maybe make `txn` 4 char name to cheat...? 😜

File: src/stores/ToastStore.ts (hunk 1 above)
Comment (jryans) [reply src/stores/ToastStore.ts:59]: But don't the two toast stores take entirely differently structured objects...?

```
                ToastStore.sharedInstance().addOrReplaceToast({
                    key: 'verifreq_' + request.channel.transactionId,
                    title: request.isSelfVerification ? _t("Self-verification request") : _t("Verification Request"),
                    icon: "verification",
                    props: {request},
                    component: sdk.getComponent("toasts.VerificationRequestToast"),
                    priority: 90,
                });
```

vs. 

```
            const ref = NonUrgentToastStore.instance.addToast(NonUrgentEchoFailureToast);
```

File: src/stores/local-echo/EchoTransaction.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/local-echo/EchoTransaction.ts:24]: they can be retried though, so it's not quite fulfilled. Success/error is probably fine enough for now, as the distinction probably doesn't matter.

File: src/stores/local-echo/GenericEchoChamber.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/local-echo/GenericEchoChamber.ts:87]: txnnnnnnn

File: src/stores/ToastStore.ts (hunk 1 above)
Comment (turt2live) [reply src/stores/ToastStore.ts:59]: `sdk.getComponent` returns the same thing as `import NonUrgentEchoFailureToast` (we're passing the class through, which is super weird)

Comment (jryans) [reply src/stores/ToastStore.ts:59]: I do not understand. One is an object, with some keys we peel off manually first. The other is a single class we pass straight into `createElement`... How can they possibly be the same?

Comment (turt2live) [reply src/stores/ToastStore.ts:59]: so the `ToastStore` takes an `IToast<C>` (which can be expanded to be `IToast<VerificationRequestToast>` for clarity), where the new one takes just the class, largely for simplicity's sake (we should probably support `props`, but tbh I don't expect this NonUrgentToast stuff to survive).

Comment (jryans) [reply src/stores/ToastStore.ts:59]: Oh oh oh, okay, I get now, thanks. 😅 😅 😅 Not sure why it took so long.

**Pull request merged**